### DISCUSSION TITLE
Support for --json_extension=GEOJSON in GCSToBigQueryOperator

### DISCUSSION THREAD

**Discussion by ModoGen** [apache/airflow#36559] [Q&A] [answered]
### Description:
I have recently started working with Apache Airflow and am engaged in a project involving the import of GeoJSON data into BigQuery. I am using the GCSToBigQueryOperator for data import but am encountering an issue with the correct interpretation of geospatial data.

### Details:
* I have a NEWLINE_DELIMITED_JSON file containing GeoJSON data. Each record in this file is separated by a newline.
* When attempting to load these data into BigQuery using the GCSToBigQueryOperator, I receive the following error: "Error while reading data, error message: JSON parsing error in row starting at position 0: JSON object specified for non-record field: geometry File ...".
* After researching, I found that the command bq load --source_format=NEWLINE_DELIMITED_JSON --json_extension=GEOJSON --autodetect {dataset}.{tablename} {mydata.geojsonl} works, as described in this article: [Using GeoJSON for BigQuery geospatial analytics](https://cloud.google.com/blog/topics/developers-practitioners/using-geojson-bigquery-geospatial-analytics?hl=en).
* I would like to implement this functionality in Apache Airflow but cannot find a way to set the --json_extension=GEOJSON option in the GCSToBigQueryOperator.

### Example:
Please note that the data examples provided in this ticket have been simplified and anonymised for data protection reasons. They only serve to illustrate the problem and do not reflect the actual data I am working with.

Example of a JSONL File:
```json
{ "type": "Feature", "properties": { "gml_id": "abc" }, "geometry": { "type": "MultiPolygon", "coordinates":  [...] } }
{ "type": "Feature", "properties": { "gml_id": "abc2" }, "geometry": { "type": "MultiPolygon", "coordinates":  [...] } }
```
Example of a Schema File:
```json
[
  {
    "name": "gml_id",
    "type": "STRING",
    "mode": "NULLABLE",
    "fields": []
  },
  {
    "name": "geometry",
    "type": "GEOGRAPHY",
    "mode": "NULLABLE",
    "fields": []
  }
]
```

### Question:
How can I configure the GCSToBigQueryOperator to use the --json_extension=GEOJSON option to properly import GeoJSON data? Is there a way to utilize or implement this functionality in Apache Airflow version v2.6.3?

### Additional Information:
Apache Airflow Version: v2.6.3
The issue arises when loading NEWLINE_DELIMITED_JSON data containing geometry objects into BigQuery.
Manual import using the bq command works, but integration into Airflow is necessary for the automation of the process.

**Comment by ModoGen** [comment]
I'm pleased to report that I was able to resolve the issue with the GCSToBigQueryOperator and the import of GeoJSON data into BigQuery. The solution involved extending the GCSToBigQueryOperator to include the necessary json_extension option for GeoJSON files. Here's how I achieved it:

Custom Operator Implementation:
I created a custom operator by extending the GCSToBigQueryOperator from the airflow.providers.google.cloud.transfers.gcs_to_bigquery module. In the custom operator, I added the json_extension parameter and set its default value to "GEOJSON". This parameter is then used to modify the job configuration in the _submit_job method, ensuring that BigQuery correctly interprets the GeoJSON data.

Here's the code for the custom operator:
```python
from airflow.providers.google.cloud.transfers.gcs_to_bigquery import GCSToBigQueryOperator

class CustomGCSToBigQueryOperator(GCSToBigQueryOperator):
    def __init__(self, *args, json_extension="GEOJSON", **kwargs):
        super().__init__(*args, **kwargs)
        self.json_extension = json_extension

    def _submit_job(self, hook, job_id: str):
        self.configuration["load"]["jsonExtension"] = self.json_extension
        return super()._submit_job(hook, job_id)
```

Results:
After implementing this custom operator, I was able to successfully import the newline-delimited JSON files containing GeoJSON data into BigQuery without encountering the previous errors. The GeoJSON data was correctly interpreted by BigQuery, which was evident from the successful parsing and loading of the geometry field in the data.

**Reply by potiuk** [reply]
Contributing it back would be a good idea :)

**Comment by raphaelauv** [comment, marked as answer]
don't use any other operator than 

`BigQueryInsertJobOperator`

for interacting with bigquery

**Reply by raphaelauv** [reply]
https://cloud.google.com/bigquery/docs/loading-data-cloud-storage-csv#sql
